### PR TITLE
sensors: move gyro filtering to sensors/vehicle_angular_velocity

### PR DESCRIPTION
**Pull request by dagar**
We currently filter all gyros (and accels), however only the primary sensor's filtered data is actually used at all. By moving the filtering downstream we can further simplify what's done in each driver and save a little bit of cpu.

### Overall changes
 - gyro filtering (low-pass and notch) only performed on primary gyro in `sensors/vehicle_angular_velocity` instead of every gyro in `PX4Gyroscope`
 - sample rate is calculated from actual updates (the fixed value was slightly wrong in many cases, and very wrong in a few)
 - In the FIFO case the array is now averaged and published in `sensor_gyro` for filtering downstream. I'll update this in the future to use the full FIFO array (if available), but right now it should be fine.
    - this also means using the notch filter with FIFO data is technically "fixed" for now (https://github.com/PX4/Firmware/pull/13874)
 - mpu6000 sample perf counter updated to only include the actual read time
 - if this is merged I'll mirror the changes for accelerometer

### CODE REVIEW

**Comment by dusan19** [comment]
What happens in case the sensor selection changes, should the filters be reset or would the switch have no significant effect?

**Comment by dagar** [comment]
> What happens in case the sensor selection changes, should the filters be reset or would the switch have no significant effect?

Good question. I didn't bother with explicitly resetting the filters because there are 2 other potential changes that will be abrupt.
1. The bias (estimated bias from ekf in sensor_bias message) is zeroed on sensor change and will take a while before it settlse for the new sensor.
    * this will be properly addressed once we're running multiple estimator instances (soon™)
    * another option could be to maintain the biases for all sensors by comparing the diff with the primary sensor + estimated bias, but I'm not sure if it's worth the effort
2. The sample rate might be different (it's not always possible to configure them identically, especially between ST and Bosch). This module measures the actual sample interval and resets the filters when the sample rate error is > 1%.

Additionally I'm also looking at adding a queue depth to `sensor_accel` and `sensor_gyro` messages so that we can do things like integrate the gyro data synchronized with optical flow measurements. Once that's in place if sensor selection changes we can do a more thorough reset and push the full set of queued samples (maybe ~ 10 messages) through the filter before publishing the new `vehicle_angular_velocity`.

File: src/lib/drivers/gyroscope/PX4Gyroscope.cpp [hunk 1]
```diff
@@ -213,45 +194,30 @@ void PX4Gyroscope::updateFIFO(const FIFOSample &sample)
...
-
-
-	// publish control data (filtered) immediately
+	// publish raw data immediately
 	{
-		bool publish_control = true;
+		// average
+		float x = (float)sum(sample.x, N) / (float)N;
+		float y = (float)sum(sample.y, N) / (float)N;
+		float z = (float)sum(sample.z, N) / (float)N;
```
Comment (jlecoeur) [inline src/lib/drivers/gyroscope/PX4Gyroscope.cpp]: How do you plan to use the full FIFO array in the future? Publish N sensor_gyro reports with fake timestamps?

File: src/modules/sensors/vehicle_angular_velocity/VehicleAngularVelocity.hpp [hunk 2]
```diff
@@ -94,12 +103,23 @@ class VehicleAngularVelocity : public ModuleParams, public px4::WorkItem
...
+
 	matrix::Dcmf _board_rotation;
 
 	matrix::Vector3f _bias{0.f, 0.f, 0.f};
 	matrix::Vector3f _offset{0.f, 0.f, 0.f};
 	matrix::Vector3f _scale{1.f, 1.f, 1.f};
 
+	hrt_abstime _last_publish{0};
+	hrt_abstime _filter_check_last{0};
+	static constexpr const float kINITIAL_RATE_HZ{1000.0f}; /**< sensor update rate used for initialization */
```
Comment (jlecoeur) [inline src/modules/sensors/vehicle_angular_velocity/VehicleAngularVelocity.hpp]: typo? INITIAL_RATE_HZ

File: src/modules/sensors/vehicle_angular_velocity/VehicleAngularVelocity.cpp [hunk 3]
```diff
@@ -77,6 +81,49 @@ void VehicleAngularVelocity::Stop()
...
+				// check if sample rate error is greater than 1%
+				if ((fabsf(_update_rate_hz - _filter_sample_rate) / _filter_sample_rate) > 0.01f) {
+					++_sample_rate_incorrect_count;
+				}
+			}
+		}
+
+		const bool sample_rate_updated = (_sample_rate_incorrect_count > 50);
+		const bool lowpass_updated = (fabsf(_lowpass_filter.get_cutoff_freq() - _param_imu_gyro_cutoff.get()) > 0.01f);
+		const bool notch_updated = (fabsf(_notch_filter.getNotchFreq() - _param_imu_gyro_nf_freq.get()) > 0.01f);
```
Comment (jlecoeur) [inline src/modules/sensors/vehicle_angular_velocity/VehicleAngularVelocity.cpp]:
```suggestion
		const bool notch_updated = ((fabsf(_notch_filter.getNotchFreq() - _param_imu_gyro_nf_freq.get()) > 0.01f)
					 || (fabsf(_notch_filter.getBandwidth() - _param_imu_gyro_nf_bw.get()) > 0.01f));
```

File: src/lib/drivers/gyroscope/PX4Gyroscope.cpp (hunk 1 above)
Comment (dagar) [reply src/lib/drivers/gyroscope/PX4Gyroscope.cpp]: I considered publishing many sensor_gyro messages, but then there's too much overhead to log effectively even as a special configuration.

At the moment I'm looking at moving all processing downstream.
 - filtering moves to vehicle_acceleration/vehicle_angular_velocity (nearly done)
 - simple calibration (offsets and scale) can be applied downstream exactly as temperature corrections are used now (WIP, but looks viable)
 - integration moves to vehicle_imu (not started yet)

Once that settles using the FIFO becomes a special case in vehicle_{acceleration, angular_velocity, imu} where the FIFO version of the sensor message is preferred if available.

If possible we might even get to the point where all sensor_{accel, gyro} consumers are aware of the FIFO alternative and then the drivers would be stripped down to only publish a single simple raw message at the sample rate.

File: src/modules/sensors/vehicle_angular_velocity/VehicleAngularVelocity.hpp (hunk 2 above)
Comment (dagar) [reply src/modules/sensors/vehicle_angular_velocity/VehicleAngularVelocity.hpp]: I don't recall, but might have been a half-assed partial attempt at following the google c++ style guide. Actually following the google c++ style guide would be `kInitialRateHz`.

Fixing...

File: src/modules/sensors/vehicle_angular_velocity/VehicleAngularVelocity.cpp (hunk 3 above)
Comment (dagar) [reply src/modules/sensors/vehicle_angular_velocity/VehicleAngularVelocity.cpp]: Thanks, fixed.

File: src/lib/drivers/gyroscope/PX4Gyroscope.cpp (hunk 1 above)
Comment (jlecoeur) [reply src/lib/drivers/gyroscope/PX4Gyroscope.cpp]: sounds good

**Pull request merged**